ipq40xx: Add support for Teltonika RUTX11

MinistrasP opened the pull request:
ipq40xx: Add support for Teltonika RUTX11

This patch adds support for Teltonika RUTX11. It's an industrial DIN rail router, with 4 ethernet
ports, 2.4G/5G dualband WiFi, Bluetooth, LTE modem with SIM switch capabilities, USB 2.0 port
and two GPIOs.

This device is very similar to RUTX10, with main difference of having a 2 SIM LTE modem.

See https://teltonika-networks.com/product/rutx11/ for more info.

Hardware:
	SoC:                 Qualcomm IPQ4018
	RAM:                 256MB DDR3
	SPI Flash 1:         XTX XT25F128B (16MB, NOR)
	SPI Flash 2:         XTX XT26G02AWS (256MB, NAND)
	Ethernet:            Built-in IPQ4018 (SoC, QCA8075), 4x 10/100/1000 ports
	WiFi 1:              Qualcomm QCA4019 IEEE 802.11b/g/n
	Wifi 2:              Qualcomm QCA4019 IEEE 802.11a/n/ac
	USB Hub:             Genesys Logic GL852GT
	Bluetooth:           Qualcomm CSR8510 (A10U)
	LED/GPIO controller: STM32F030 with custom firmware
	Buttons:             Reset button
	Leds:                Power (green, cannot be controlled)
                             WiFi 2.4G activity (green)
                             WiFi 5G activity (green)
                             SIM 1 insert status(green)
		             SIM 2 insert status(green)
			     WWAN WiFi (green)
			     WWAN Ethernet (green)
			     Mobile 3g (green)
			     Mobile 4g (green)
		             Mobile level 1 (green)
			     Mobile level 2 (green)
			     Mobile level 3 (green)
			     Mobile level 4 (green)
			     Mobile level 5 (green)

LED's and LTE controlling GPIO's are driven by external STM32 chip, which needs a separate kernel
driver. It can be extracted from official Teltonika GPL sources, or from a following feed:
https://github.com/0xFelix/teltonika-rutx-openwrt

Software can be flashed two ways - OEM web interface or u-boot web recovery.

Flashing with OEM web interface:
1. Log into the router web interface (stock address is 192.168.1.1)
2. Navigate to System->Firmware
3. Set "Keep settings" to off
4. Upload openwrt-ipq40xx-generic-teltonika_rutx11-squashfs-nand-factory.ubi
5. Wait until upgrade completes

Flashing with u-boot web recovery:
1. Set PC static ip to 192.168.1.2, gateway to 192.168.1.1
2. Remove power cable and push the reset button
3. Insert power cable with reset button still pressed
4. After 5 seconds release the button and open web interface at 192.168.1.1
(Sometimes cache persists, so open private browsing)
5. Upload openwrt-ipq40xx-generic-teltonika_rutx11-squashfs-nand-factory.ubi
6. Wait until upgrade completes

Note: Mobile is work in progress.

Thanks for your contribution to OpenWrt!

To help keep the codebase consistent and readable,
and to help people review your contribution,
we ask you to follow the rules you find in the wiki at this link
https://openwrt.org/submitting-patches

Signed-off-by: Kasparas Elzbutas <elzkas@gmail.com>

PolynomialDivision commented:
Please, rework your commit message. Signed off has to be at the end of the commit message and not in the commit header (first line). Further, it needs to be your real name and not any github nick.

Further remove:
```
 Changes to be committed:
	modified:   feeds.conf.default
	modified:   target/linux/ipq40xx/base-files/etc/board.d/01_leds
	modified:   target/linux/ipq40xx/base-files/etc/board.d/02_network
	modified:   target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata
	new file:   target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx11.dts
	modified:   target/linux/ipq40xx/image/generic.mk
	modified:   target/linux/ipq40xx/patches-5.10/901-arm-boot-add-dts-files.patch
	modified:   target/linux/ipq40xx/patches-5.15/901-arm-boot-add-dts-files.patch
```

Further, add the soc name to the heading like: `ipq40xx: ...`

Please, also do a DSA conversion:
https://github.com/openwrt/openwrt/pull/4721

MinistrasP commented:
Seems like the first port of QCA8075 isn't connected, so the numbering starts from 2 up to 5. Where 2,3,4 are LAN ports and 4 is a WAN port. Tested on both RUTX10 and RUTX11 which are identical in that regard.

MinistrasP commented:
A question - LTE modem needs an internal restart with AT+CFUN=0 and AT+CFUN=1 everytime SIM gets removed/inserted. Should i add some kind of script in the merge, or should i leave it for the user?

0xFelix commented:
> Seems like the first port of QCA8075 isn't connected, so the numbering starts from 2 up to 5. Where 2,3,4 are LAN ports and 5 is a WAN port. Tested on both RUTX10 and RUTX11 which are identical in that regard.

Can you do the DSA conversion for the RUTX10 too? That would be great.

MinistrasP commented:
> > Seems like the first port of QCA8075 isn't connected, so the numbering starts from 2 up to 5. Where 2,3,4 are LAN ports and 5 is a WAN port. Tested on both RUTX10 and RUTX11 which are identical in that regard.
> 
> Can you do the DSA conversion for the RUTX10 too? That would be great.

Sure, will do it later.

hauke commented on feeds.conf.default:
```diff
@@ -6,3 +6,4 @@ src-git-full telephony https://git.openwrt.org/feed/telephony.git
 #src-git-full targets https://github.com/openwrt/targets.git
 #src-git-full oldpackages http://git.openwrt.org/packages.git
 #src-link custom /usr/src/openwrt/custom-feed
+src-git teltonika https://github.com/0xFelix/teltonika-rutx-openwrt.git
```
Please remove this feed

hauke commented on target/linux/ipq40xx/base-files/etc/board.d/02_network:
```diff
@@ -37,11 +37,14 @@ ipq40xx_setup_interfaces()
 		;;
 	aruba,ap-303h|\
-	teltonika,rutx10)
+	teltonika,rutx10|\
```
This looks wrong.

MinistrasP commented:
> Please remove the merge and squash this together into one commit

You mean the merge with master branch, right?

csharper2005 commented:
> You mean the merge with master branch, right?

https://openwrt.org/docs/guide-developer/working-with-github-pr#alternative_squashing_advice

MinistrasP commented:
I think i have done everything.

MinistrasP commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx11.dts:
```diff
@@ -0,0 +1,234 @@
+	status = "okay";
+};
+
+&swport2 {
```
Okay i changed it back and tested, now everything compiles and works.

markmentovai commented on target/linux/ipq40xx/base-files/etc/board.d/01_leds:
```diff
@@ -86,6 +86,10 @@ netgear,ex6150v2)
 	ucidef_set_led_wlan "wlan2g" "WLAN2G" "green:router" "phy0tpt"
 	ucidef_set_led_wlan "wlan5g" "WLAN5G" "green:client" "phy1tpt"
 	;;
+teltonika,rutx11)
```
Can this be handled directly in the `.dts` file (as it is for teltonika,rutx10)?

MinistrasP replied:
Fixed, testing.

markmentovai commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx11.dts:
```diff
@@ -0,0 +1,210 @@
+			compatible = "gpio-leds";
+
+			led_wifi2g {
+				label = "wifi_2g";
```
LED labels normally follow the pattern color:function. This would be `green:wlan2g` and line 140 would be `green:wlan5g`. Similar for the other LEDs that follow.

MinistrasP replied:
Fixed

markmentovai commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx11.dts:
```diff
@@ -0,0 +1,210 @@
+			};
+
+			led_ssid_1 {
+				label = "ssid_1";
```
What are these 5 “SSID” LEDs about?

<img src="https://wiki.teltonika-networks.com/wikibase/images/c/ca/Rutx11_hd_2.png"/>

Those look like they’re intended to indicate RSSI.

`mikrotik,sxtsq-5-ac` uses `green:rssi{low,mediumlow,medium,mediumhigh,high}` for this, although I can see how something numeric might be more attractive.

MinistrasP replied:
Replaced all SSID labels with correct ones.

markmentovai commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx11.dts:
```diff
@@ -0,0 +1,210 @@
+			led_ssid_1 {
+				label = "ssid_1";
+				gpios = <&stm32_io 7 GPIO_ACTIVE_HIGH>;
+				linux,default-trigger = "timer";
```
What’s this for?

Same for the other 4 “ssid” but probably really RSSI LEDs.

MinistrasP replied:
Replaced all SSID labels with correct ones.

markmentovai replied:
This question wasn’t about the “SSID” name, it was about this specific line:

```
				linux,default-trigger = "timer";
```

> What’s this for?

markmentovai replied:
There’s a standard way to use LED to indicate boot status in OpenWrt, via `/etc/diag.sh`. You can set some `led-*` aliases in the `.dts` file to enable this behavior (`led-boot`, `led-failsafe`, `led-running`, `led-upgrade`). Have a look at some of the neighboring `.dts` files to see how this works.

MinistrasP replied:
Added boot behaviour for rssi low led.

markmentovai commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx11.dts:
```diff
@@ -0,0 +1,210 @@
+	model = "Teltonika RUTX11";
+	compatible = "teltonika,rutx11";
+
+	memory {
```
This is already in `qcom-ipq4018-rutx.dtsi`.

MinistrasP replied:
Removed it.

simonastamosaitis commented on target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -127,6 +127,10 @@ case "$FIRMWARE" in
 		caldata_extract_mmc "0:ART" 0x1000 0x2f20
 		ath10k_patch_mac $(mmc_get_mac_binary ARTMTD 0x0)
 		;;
+	teltonika,rutx11)
```
New wireless configuration in DTS doesn't work in RUTX10. It should be registered as issue.
U-boot changes DTS partitions like in this commit: https://github.com/openwrt/openwrt/commit/af425e42212d476dd95ec31f18d85b19004c9268

MinistrasP replied:
As simonastamosaitis said wireless configuration in DTS doesn't work on rutx10 - I tested it a few weeks ago on both RUTX10 and on RUTX11. Both of them can't get correct pre-cal's.

MinistrasP replied:
Sure.

MinistrasP commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx11.dts:
```diff
@@ -0,0 +1,210 @@
+		reg = <0x80000000 0x10000000>;
+	};
+	
+	aliases {
```
Removed it

MinistrasP commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx11.dts:
```diff
@@ -0,0 +1,210 @@
+			pinctrl-names = "default";
+			phy-reset-gpio = <&tlmm 62 0>;
+
+			ethphy4: ethernet-phy@4 {
```
Removed it

MinistrasP commented:
Fixed and tested most of the stuff @markmentovai, except wi-fi calibration - new one doesn't work, I am leaving it as it is.

MinistrasP commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx11.dts:
```diff
@@ -0,0 +1,139 @@
+	compatible = "teltonika,rutx11";
+	
+	aliases {
+		led-boot = &led_rssi;
```
I'll do it. 

> Is there no software control of the power LED?
No, it's connected directly to power.

MinistrasP committed:
ipq40xx: Add support for Teltonika RUTX11

This patch adds support for Teltonika RUTX11. It's an industrial DIN rail router, with 4 ethernet
ports, 2.4G/5G dualband WiFi, Bluetooth, LTE modem with SIM switch capabilities, USB 2.0 port
and two GPIOs.

See https://teltonika-networks.com/product/rutx11/ for more info.

Hardware:

SoC: Qualcomm IPQ4018
RAM: 256MB DDR3
SPI Flash 1: XTX XT25F128B (16MB, NOR)
SPI Flash 2: XTX XT26G02AWS (256MB, NAND)
Ethernet: Built-in IPQ4018 (SoC, QCA8075), 4x 10/100/1000 ports
WiFi 1: Qualcomm QCA4019 IEEE 802.11b/g/n
Wifi 2: Qualcomm QCA4019 IEEE 802.11a/n/ac
USB Hub: Genesys Logic GL852GT
Bluetooth: Qualcomm CSR8510 (A10U)
LED/GPIO controller: STM32F030 with custom firmware
Buttons: Reset button
Leds: Power (green, cannot be controlled)
WiFi 2.4G activity (green)
WiFi 5G activity (green)
SIM 1 insert status(green)
SIM 2 insert status(green)
WWAN WiFi (green)
WWAN Ethernet (green)
Mobile 3g (green)
Mobile 4g (green)
Mobile level 1 (green)
Mobile level 2 (green)
Mobile level 3 (green)
Mobile level 4 (green)
Mobile level 5 (green)

LED's and LTE controlling GPIO's are driven by external STM32 chip, which needs a separate kernel
driver. It can be extracted from official Teltonika GPL sources, or from a following feed:
https://github.com/0xFelix/teltonika-rutx-openwrt

Software can be flashed two ways - OEM web interface or u-boot web recovery.

Flashing with OEM web interface:
	1. Log into the router web interface (stock address is 192.168.1.1)
	2. Navigate to System->Firmware
	3. Set "Keep settings" to off
	4. Upload openwrt-ipq40xx-generic-teltonika_rutx11-squashfs-nand-factory.ubi
	5. Wait until upgrade completes

Flashing with u-boot web recovery:
        1. Set PC static ip to 192.168.1.2, gateway to 192.168.1.1
        2. Remove power cable and push the reset button
        3. Insert power cable with reset button still pressed
        4. After 5 seconds release the button and open web interface at 192.168.1.1
        5. Upload openwrt-ipq40xx-generic-teltonika_rutx11-squashfs-nand-factory.ubi
        6. Wait until upgrade completes

Note: Currently WWAN support is not available - it is up to the user to install
the required software for built in LTE modem. Kernel driver for external STM32 chip
(mentioned above) is required to enable LTE modem.

Signed-off-by: Kasparas Elzbutas <elzkas@gmail.com>

dotSynergy commented:
referencing this post: https://forum.openwrt.org/t/teltonika-rutx-dealing-with-uboot-selected-multiple-device-trees/165423/4
i have some updates on device support.
Teltonika switched from STM32 to shiftreg for GPIO/LEDs, rendering RUTX10 support and this pr partially obsolete.
if this pr is still alive, i can take over and complete support, as i have adapted scripts that build a single image with multiple dtses for both STM32 and shiftreg version.

0xFelix commented:
@dotSynergy It would be great if you could pick up the work on this. Unfortunately I can't help much because I don't have a device to test anymore.

0xFelix commented:
> sure. since it affects RUTX10 too, would the fix for supporting multiple dtses warrant a different pr?

Don't the DTSes of these devices share a `.dtsi`? So IMO it could still be in one PR.

0xFelix commented:
Ah I think I misunderstood, you wanted to have one PR for the multiple DTSes and one for DSA? That makes sense.

dotSynergy commented:
> Ah I think I misunderstood, you wanted to have one PR for the multiple DTSes and one for DSA? That makes sense.

yes, that would be my intention. i already built a pr for multiple DTSes, but honestly i wouldn't know where to start for the DSA part

0xFelix commented:
@dotSynergy AFAIR only some slight changes to the DTS and the default network config are required. IIRC the hardest part is getting the assignment of ports right with the labels on the device. Here are some examples: https://github.com/openwrt/openwrt/pull/12478, https://github.com/openwrt/openwrt/pull/12477

dotSynergy commented:
> @dotSynergy AFAIR only some slight changes to the DTS and the default network config are required. IIRC the hardest part is getting the assignment of ports right to match the labels on the device. Here are some examples: #12478, #12477

thanks for the pointers. i can try that on RUTX11, but unfortunately i don't have a RUTX10 to test this on. i  might still propose the impl for RUTX11 though, soon.
